zcash_client_backend/src/fees/common.rs

```diff
@@ -400,91 +400,78 @@ where
             .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?,
     );
 
-    let total_out_plus_fee_without_change =
-        (total_out + fee_without_change).ok_or_else(overflow)?;
-    let total_out_plus_fee_with_change = (total_out + fee_with_change).ok_or_else(overflow)?;
+    let total_out_plus_min_fee = (total_out + min_fee).ok_or_else(overflow)?;
+    let total_out_plus_max_fee = (total_out + max_fee).ok_or_else(overflow)?;
 
     #[allow(unused_mut)]
     let (mut change, fee) = {
-        if transparent && total_in < total_out_plus_fee_without_change {
-            // Case 1 for a tx with all transparent flows.
+        if total_in < total_out_plus_min_fee {
+            // Insufficient inputs exist to pay the minimum fee; there's no way we can construct
+            // the transaction.
             return Err(ChangeError::InsufficientFunds {
                 available: total_in,
-                required: total_out_plus_fee_without_change,
-            });
-        } else if transparent && total_in == total_out_plus_fee_without_change {
-            // Case 2 for a tx with all transparent flows.
-            (vec![], fee_without_change)
-        } else if total_in < total_out_plus_fee_with_change {
```

Comment:
This line was the primary cause of the error: we were computing the maximum fee, then checking that the total of the inputs was sufficient to satisfy that maximum fee plus the output total; however, the maximum fee potentially takes into account change outputs that are never actually created if there's insufficient change to perform splitting for note management purposes.

Reply:
Yes, because there was no splitting when this code was written. So there was at most one change note and it was correct for this case to be an error, because neither adding nor omitting the change note would work. The bug was introduced in #1579, and we failed to catch it in review.